WIP: Custom backends for Dynamic Selection (#2220)
uxlfoundation/oneDPL: rfcs/proposed/dynamic_selection_customization/README.md

a manditory template argument and this can be used to set the resource type in the backend.

Is it therefore becomes unnecessary to explicitly provide the resource type in the default backend.
For example, in the code below a `round_robin_policy` is constructed by passing a vector of
pointers to `tbb::task_group`.  The type of the resource `tbb::task_group *` can be deduce and
passed as a template argument to the default backend. 

```cpp
    tbb::task_group t1, t2;
    ex::round_robin_policy p{ { &t1, &t2 } };
```

If the policy is default constructed or constructed for deferred initialization, the
resource type must be given explicitly.

```cpp
    ex::round_robin_policy<tbb::task_group*> p1{ };
    ex::round_robin_policy<tbb::task_group*> p2{ deferred_initialization_t };
```
#### Default `wait_t<T>`

The `wait_t<T>` can only be deduced at the time a user function is passed to submit since
it is the type returned by the user's function. There is no way to generally determine a
`wait_t<T>` from an arbitrary `resource_t<T>` and therefore the default backend cannot
provide a meaningful `wait_t<T>`.

> 

Reviewer: Should we propose in this RFC to remove `wait_t<T>` altogether?

Author: I think so, yes.

#### Default `b.submit(s, f, args…)`

The main role that `submit` plays in a backend is to add instrumentation around the call to `f`.
Experience has shown us that most backends perform four basic steps in their implementation
of the `submit` function:

1. Do any setup needed for implementing reporting before calling `f`.
2. The function `f` is called and it return value captured.
3. Do any setup needed for implementing reporting after calling `f`, perhaps using what was returned by `f`.

> 4. The *Submission* object, which typically wraps what is returned by `f`, is constructed and returned.

Reviewer: nitpick: 
This seems to go back and forth between present tense and past tense, we should keep the same orientation for all steps

It is not possible for a default backend to properly instrument execution for an unkown resource type to provide

> reporting of `task_time`, `task_submission` and `task_completion`. And so a default implementation
> cannot provide useful implementations for step 1 and 3. However, it can provide more finer-grained hooks

Reviewer: ```suggestion
reporting of `task_time`, `task_submission` and `task_completion`. Therefore a default implementation
cannot provide useful implementations for step 1 and 3. However, it can provide more fine-grained hooks
```

that can be overridden to provide steps 1 and 3, without requiring a custom backend to reimplemnted the entire
four step pattern.

We propose that the default backend provide a `submit` function that implements the four step pattern

> but also calls `instrument_before_impl` and `instrument_after_impl` functions that can be overridden

Reviewer: ```suggestion
but also calls `instrument_before_impl` and `instrument_after_impl` functions that can be individually overridden
```

to add instrumentation by backends that need it.

```cpp
    template <typename SelectionHandle, typename Function, typename... Args>
    auto submit(SelectionHandle s, Function&& f, Args&&... args) {
        instrument_before_impl(s); // do insrumentation before calling `f`

        return instrument_after_impl( // do insrumentation before calling `f`, and create Submission type
            s,

>             std::forward<Function>(f)(unwrap(s), std::forward<Args>(args)...) // call f

Reviewer: nitpick:
for explanation, rearrange code to allow function call to appear above `instrument_after_impl`

        );
    }
```

#### Default *Submission* type

> A *Submission* object must support `s.wait()` and `s.unwrap`. A default backend cannot meaningfully
> implement `wait` on an arbitrary type, but can wrap a type, such as `sycl::queue` or `tbb::task_group`
> that provides a `wait` member function itself and call that `wait` member function if it finds it.

Reviewer: ```suggestion
A *Submission* object must support `s.wait()` and `s.unwrap`. A default backend cannot meaningfully
implement `wait` on an arbitrary type, but it can wrap a type, such as `sycl::queue` or `tbb::task_group`
that provides a `wait` member function itself and call that `wait` member function if it exists.
```

> `s.unwrap` can return what was return by the user's function `f`.

Reviewer: ```suggestion
`s.unwrap` can return what was returned by the user's function `f`.
```

                        });
    ex::wait(s);
```

In the above code, no special backend is written for `tbb::task_group*`, but
a policy can be created with a given set of resources that is can be used to submit
and wait on a submission.

#### Default `b.get_resources()`

The `get_resources` function returns the set of resources managed by the backend. If a policy is
constructed with a `std::vector<T>`, then `resource_t<Backend>` is `T` and the function
`get_resources` can return the vector that was passed to the policy constructor, which in turn,
is used to construct the backend.

```cpp
    tbb::task_group t1, t2;

    // constructs backend with std::vector<tbb::task_group*>{ &t1, &t2 }
    ex::round_robin_policy p{ { &t1, &t2 } };

    // returns a std::vector<tbb::task_group*>{ &t1, &t2 }
    // because backend returns the vector it was constructed with
     auto v = p.get_resources();
```

However, a default constructed policy has no information about what set makes sense for
an arbitrary resource type. And so `get_resources` will return an empty vector.

```cpp
    ex::round_robin_policy<tbb::task_group*> p1{ };

    // v is empty
     auto v = p.get_resources();

> ```

Reviewer: mention deferred initialization.

The function `get_submission_group()` is used by a policy's `get_submission_group()` function
to return an object that has a member function `void wait()`. A call to `wait` blocks until all
incomplete submissions are done.

There is no wait for a default backend to implement a type that waits for all submission to a set
of arbitrary resource types. There are several possible alternatives including:

* Return a submission group that has an empty `wait` function.
* Return a submission group that throws an exception if it is waited on.
* Calls to `get_submission_group` fail to compile when the default backend is used.

This proposal recommends the last suggestion.

```cpp
    tbb::task_group t1, t2;

    // constructs backend with std::vector<tbb::task_group*>{ &t1, &t2 }
    ex::round_robin_policy p{ { &t1, &t2 } };

    auto g = p.get_submission_group(); // fails to compile
```

#### Default `lazy_report_v<T>`

The default backend, which does not provide any reporting support at all, will
have `lazy_report_v<T>` as `false`.

## How to customize for specific resource types

We proposed using a CRTP (Curiously Recurring Template Pattern) to define backends since this

> will allow developers to flexibly mix-in functionaility from the base. We decided against free

Reviewer: ```suggestion
will allow developers to flexibly mix-in functionality from the base. We decided against free
```

                                          std::forward<Args>(args)...)
            );
        }

        template <typename SelectionHandle>
        void instrument_before_impl(SelectionHandle s) { }

        template <typename SelectionHandle, typename WaitType>
        auto instrument_after_impl(SelectionHandle s, WaitType w) {
            return default_submission{w};
        }

        std::vector<resource_type> get_resources_impl() const noexcept {
            return resources_;
        }

    };
```

We also propose to provide a catch-all [default backend](#default-backend) for uncustomized resource
types:

```cpp
    // A general default backend for ResourceType
    template< typename ResourceType >
    class default_backend : public backend_base<ResourceType, default_backend<ResourceType>> {
    public:
        using resource_type = ResourceType;
    };
```

A backend that relies on the defaults but overrides some functionality can then specialize the
default backend for the type. For example, the existing SYCL backend can be implemented as:

```cpp
template< >
class default_backend<sycl::queue> : public backend_base<sycl::queue, default_backend<sycl::queue>>
{

>   //* override as needed

Reviewer: Is this a TODO for the RFC?
Do we want to describe what is actually necessary to define `sycl::queue` backend here?

In the PoC as it stands, it actually its ~30 lines longer implementation than the previous version... can we cut it down more?

Or is there a different simpler example we should use here?

Author: I will add different examples.

};
```

And the use of this SYCL backend by a policy, such as `round_robin_policy` is implemented by a policy
as:

```cpp

> template <typename ResourceType = sycl::queue, typename Backend = default_backend<sycl::queue>>

Reviewer: ```suggestion
template <typename ResourceType = sycl::queue, typename Backend = default_backend<ResourceType>>
```

# Simplified Customization of Backends and Policies for Dynamic Selection

Dynamic Selection is a Technology Preview feature 
[documented in the oneDPL Developer Guide](https://www.intel.com/content/www/us/en/docs/onedpl/developer-guide/2022-8/dynamic-selection-api.html)
and its current design is described by 
[an experimental RFC](https://github.com/uxlfoundation/oneDPL/tree/main/rfcs/experimental/dynamic_selection).
When applying Dynamic Selection to use cases, there is often the desire to add a new
backend or new selection policy. While it is possible to add new policies or backends
by following the contract described in the current design, the process is non-trivial

> and requires some (unecessarily) verbose code.

Reviewer 2: typo? : unecessarily ->unnecessarily

Author: Fixed in coming re-write.

| `get_resources()` | Returns a `std::vector<resource_t<T>>`. |
| `get_submission_group()` | Returns an object that has a member function `void wait()`. Calling this `wait` function blocks until all previous submissions to this backend are complete. |

## Use Cases to Improve for Backends

| Use Case | Description |
| --------------------- | -------- |
| Default Backend | No custom backend at all. |
| Uninstrumented Backend | Backend overrides a small part of default backend: `wait_type`, `submission_group`, for better waiting support. |
| Instrumented Backend | Backend includes instrumentation for reporting |

The requirements of *Submission* and *Selection* objects are provided in
[the current design description](https://github.com/uxlfoundation/oneDPL/tree/main/rfcs/experimental/dynamic_selection).

### Default Backend

For policies that require no reporting of `execution_info`, it should be simple to use a new resource type,
perhaps even without writing a custom backend. Below is an example that uses round-robin to rotate through

> through pointers to queues.

Reviewer 2: repeated word "through"

Author: Will do.

```cpp
    sycl::queue q1, q2;
    ex::round_robin_policy p{ { &q1, &q2 } };

    auto s1 = onedpl::dpl::experimental::submit(p, [](sycl::queue *qp) -> sycl::queue {
        qp->submit(/*...*/);
        return *q;
    }

    auto s2 = onedpl::dpl::experimental::submit(p, [](sycl::queue *qp) -> sycl::queue {
        qp->submit(/*...*/);
        return *q;
    }

    wait(s1);
    wait(s2);
```

#### submit

The main role that `submit` plays in a backend is to add instrumentation around the call to `f`.
Experience has shown us that most backends perform four basic steps in their implementation
of the `submit` function:

1. Do any setup needed for implementing reporting before calling `f`.
2. The function `f` is called.
3. Do any setup needed for implementing reporting after calling `f`, perhaps using what was returned by `f`.
4. The *Submission* object, which typically wraps what is returned by `f`, is constructed and returned.

If a backend writer doesn't care if their backend works with policies that require reporting, then
they can use a default implementation of submit that simply calls `f` and returns the backends *Submission*

> object. 

Reviewer 2: We could add explicitly here that, in this case steps 1 and 3 are not needed. Which is part of the simplification we seek to achieve.

Author: Will do

A *Submission* object must support `s.wait()` and `s.unwrap`. A default backend cannot meaningfully
implement `wait` on an arbitrary type, but could easily wrap a type, such as `sycl::queue` or `tbb::task_group`
that provides a `wait` member function and call that `wait` member function if it finds it, and in other cases
provide an empty `s.wait` function. `s.unwrap` usually returns what was returned by the user's function `f`.
Our simple example above returns a `sycl::queue` from its function and so a default backend can easily synthesize
a meaningful *Submission* type from it, with `s.wait` calling `q.wait()` and `s.unwrap` returning the queue.

#### get_resources

The `get_resources` function can default to returning the vector that is passed to the policy, and
then down to the backend's constructor. A policy that is created with a default constructor will

> have no resources in this case.

Reviewer 2: We could maybe add that: it is therefore imperative that a user either passes the list of resources at construction of the policy or call initialize to set the resources.

Author: Hopefully addressed in rewrite that I'll post soon.

#### get_submission_group

The `get_submission_group` is not easily implemented in a meaningful way. It must return a type that
defines a member function `wait`. But since there is no way to know how to wait on all previous submissions
for an arbitrary backend resource, this function will likely need to return a dummy type or 

> `get_submission_group` be undefined for the default backend.

Author: Lets consider if there's a meaningful default or not. I'm not sure that there is.

    // and default_backend<sycl::queue> inherits from backend_base<sycl::queue, default_backend<sycl::queue>>
    round_robin_policy p;
```

A third-party can create a custom backend by using the `backend_base`:

```cpp
namespace third_party {
    class custom_resource {
        // implementation
    };

    class custom_backend : public oneapi::dpl::experimental::backend_base<custom_resource, custom_backend> {
        // override _impl functions as desired
    };
}
```

A round-robin policy using such a custom resource and policy would be define as shown:

```cpp
    round_robin_policy<third_party::custom_resource, third_party::custom_backend> p;
```

We propose adding a trait `backend_for_resource` that is used by policies to find the
correct backend for a given resource type, as shown below:

```cpp
namespace oneapi {
namespace dpl {
namespace experimental {

template <typename ResourceType>
struct backend_for_resource
{
    using backend_t = default_backend<ResourceType>;
};

template <typename ResourceType = sycl::queue, typename Backend = backend_for_resource<ResourceType>::backend_t>
struct round_robin_policy
{
    // policy implementation not shown here
};

> }
> }
> }

Reviewer: ```suggestion
}}} //namespace oneapi::dpl::experimental
```

```

There would be three ways then that a custom backend can be found:

1. The backend writer can add a specialized `backend_for_resource` in the `experimental::dpl` namespace.
2. The backend writer can specialized `default_backend` in the `experimental::dpl` namespace.
3. The backend can be provide explicitly to the policy.

The code below demonstrates each of these approaches:

```cpp
namespace third_party {

    // 1. specialize backend_for_resource in oneapi::dpl::experimental
    class custom_resource_1 {
        // implementation
    };

    class custom_backend_1 : public oneapi::dpl::experimental::backend_base<custom_resource_1, custom_backend_1> {
        // override _impl functions as desired
    };

    // 2. specialize default_backend in oneapi::dpl::experimental
    class custom_resource_2 {
        // implementation
    };

    // 3. no specializations, only available explicitly
    class custom_resource_3 {
        // implementation
    };

    class custom_backend_3 : public oneapi::dpl::experimental::backend_base<custom_resource_3, custom_backend_3> {
        // override _impl functions as desired
    };

> }

Reviewer: ```suggestion
} //namespace third_party
```

namespace oneapi {
namespace dpl {
namespace experimental {

    // 1. specialize backend_for_resource in oneapi::dpl::experimental
    template <>
    struct backend_for_resource<third_party::custom_resource_1>
    {
        using backend_t = third_party::custom_backend;
    };

    // 2. specialize default_backend in oneapi::dpl::experimental
    template< >
    class default_backend<third_party::custom_resource_2> 
        : public backend_base<third_party::custom_resource_2, default_backend<third_party::custom_resource_2>> {
    public:
        using resource_type = third_party::custom_resource_2;
        // override _impl functions as desired
    };

> }
> }
> }

Reviewer: ```suggestion
}}} //namespace oneapi::dpl::experimental
```

    // create pairs of arenas and task_groups, one per numa node
    std::vector<tbb::numa_node_id> numa_nodes = tbb::info::numa_nodes();
    std::vector<pair_t> pairs;

    for (int i = 0; i < numa_nodes.size(); i++) {
        pairs.emplace_back(pair_t(new tbb::task_arena{tbb::task_arena::constraints(numa_nodes[i]), 0},
                               new tbb::task_group{}) );
    }
    // end creating default arenas and groups

    ex::round_robin_policy<pair_t> rr{ pairs };

    // helper struct for waiting on the work in the pair
    struct WaitType {
        pair_t pair;
        void wait() { pair.first->execute([this]() { pair.second->wait(); }); }
    };
    std::vector<WaitType> submissions;

    for (auto i : numa_nodes) {
        auto w = ex::submit( rr, 
                             [](pair_t p) {
                                p.first->enqueue(p.second->defer([]() { std::printf("o\n"); }));
                                return WaitType{ p };
                             }
                            );
        submissions.emplace_back(w.unwrap());
    }

    for (auto& s : submissions)
        ex::wait(s);
}
```

> In this example, we can see that the user has to manual construct a vector of

Reviewer: ```suggestion
In this example, we can see that the user has to manually construct a vector of
```
